lib/rucio/tests/test_tpc.py

```diff
@@ -0,0 +1,168 @@
+def test_tpc(rse1, rse2, root_account, test_scope, did_factory, rse_client, rule_client, artifact):
+    if rse1 is None or rse2 is None:
+        pytest.skip("TPC tests need at least 2 containerized rse's for execution}")
+    base_file_name = generate_uuid()
+    test_file = did_factory.upload_test_file(rse1['rse_name'], name=base_file_name + '.000', return_full_item=True)
+    test_file_did_str = '%s:%s' % (test_file['did_scope'], test_file['did_name'])
+    test_file_did = {
+        'scope': test_scope,
+        'name': test_file['did_name']
+    }
+    test_file_name_hash = hashlib.md5(test_file_did_str.encode('utf-8')).hexdigest()
+    test_file_expected_pfn = '%s/%s/%s/%s' % (test_file_did['scope'], test_file_name_hash[0:2], test_file_name_hash[2:4], test_file_did['name'])
+
+    rse1_hostname = rse_client.get_protocols(rse1['rse_name'])[0]['hostname']
+    rse2_hostname = rse_client.get_protocols(rse2['rse_name'])[0]['hostname']
+
+    rule_id = add_rule(dids=[test_file_did], account=root_account, copies=1, rse_expression=rse2['rse_name'],
+                       grouping='NONE', weight=None, lifetime=None, locked=False, subscription_id=None)
+    rule = rule_client.get_replication_rule(rule_id[0])
+
+    re_evaluator(once=True)
+
+    assert rule['locks_ok_cnt'] == 0
+    assert rule['locks_replicating_cnt'] == 1
+
+    transfer_requestss = get_transfer_requests_and_source_replicas(rses=[rse1['rse_id'], rse2['rse_id']])
+    for transfer_requests in transfer_requestss:
+        for transfer_request in transfer_requests:
+            if transfer_requests[transfer_request]['rule_id'] == rule_id[0]:
+                src_url = transfer_requests[transfer_request]['sources'][0][1]
+                dest_url = transfer_requests[transfer_request]['dest_urls'][0]
+                check_url(src_url, rse1_hostname, test_file_expected_pfn)
+                check_url(dest_url, rse2_hostname, test_file_expected_pfn)
+
+    # Run Submitter
+    submitter.run(once=True)
+
+    # Get FTS transfer job info
+    fts_transfer_id, fts_transfer_status = list_fts_transfer(timeout=40)
+
+    # Check FTS transfer job
+    assert fts_transfer_id is not None and fts_transfer_status is not None
+    assert fts_transfer_status in ['SUBMITTED', 'ACTIVE']
+
+    fts_transfer_status = poll_fts_transfer_status(fts_transfer_id)
+    assert fts_transfer_status == 'FINISHED'
+
+    poller.run(once=True, older_than=0)
+    finisher.run(once=True)
+    rule = rule_client.get_replication_rule(rule_id[0])
+    assert rule['locks_ok_cnt'] == 1
+    assert rule['locks_replicating_cnt'] == 0
+
+    if artifact is not None:
+        date = datetime.date.today().strftime("%Y-%m-%d")
+        with open(artifact, 'w') as artifact_file:
+            artifact_file.write(
+                f"/var/log/fts3/{date}/{rse1['rse_name'].lower()}__{rse2['rse_name'].lower()}/*__{fts_transfer_id}"
+            )
```

Comment:
this seems a bit weird, since it builds a file name with a date (which could be wrong) and just assumes that the file is there. I think there should at least be an existence check here.

Reply:
Hey @bziemons, this piece of code is executed inside the dev_rucio container. It writes the location of the expected fts transfer log file for the transfer created by the test in the artifact i.e. /tmp/test_tpc.artifact in this case. This artifact is used in the 'Verify TPC transfers..' step of the workflow. If the file is not present, that step will fail:
https://github.com/rucio/rucio/pull/4569/files#diff-d917339141f036bb9e9a987412f09fc4c6780f8c0cd5ec1d89c4652c90032127R91

I agree that the date could be wrong, however, since the host machine of fts and rucio containers is the same, they would get the same value for date (unless the container timezones are configured differently, which is not the case as per their Dockerfile's). FTS uses a directory named ${date} to store the transfer logs. If FTS changes the directory structure for its logs, the test would fail as well. Open to suggestions for handling this. :)

Reply:
I get that, but I would suggest failing as early as possible which also helps debugging problems that may arise here; translating assumptions to asserts basically. The following would be my proposal:

```py
date = datetime.date.today().strftime("%Y-%m-%d")
fts_log_file = f"/var/log/fts3/{date}/{rse1_name.lower()}__{rse2_name.lower()}/*__{fts_transfer_id}"
assert os.path.exists(fts_log_file), 'FTS log file was not found'
with open(artifact, 'w') as artifact_file:
    artifact_file.write(fts_log_file)
```

Reply:
Hey @bziemons, thanks for the comments :). I think I need to elaborate since the suggested change
```
assert os.path.exists(fts_log_file), 'FTS log file was not found'
```

will always fail as the `fts_log_file` does not exist inside the `dev_rucio_1` container.  

```
[root@rucio-devmaany rucio]# docker exec -it dev_rucio_1 bash
[root@dc95631ee443 log]# ls -la /var/log/
total 368
drwxr-xr-x 1 root root     32 Apr 28 07:49 .
drwxr-xr-x 1 root root     28 Apr 28 07:47 ..
-rw------- 1 root utmp      0 Nov 13  2020 btmp
-rw-r--r-- 1 root root    193 Nov 13  2020 grubby_prune_debug
drwx------ 1 root root     41 May 27 07:50 httpd
-rw-r--r-- 1 root root 292000 Apr 28 07:47 lastlog
drwxrwxrwx 1 root root     66 May 27 07:50 rucio
-rw------- 1 root root  64000 Apr 28 07:47 tallylog
-rw-rw-r-- 1 root utmp      0 Nov 13  2020 wtmp
-rw------- 1 root root  12235 Apr 28 07:47 yum.log
```

It only exists inside the `dev_fts_1` container. 

```
[root@rucio-devmaany rucio]# docker exec -it dev_fts_1 bash
[root@dc95631ee443 /]# ls -la /var/log/
total 372
drwxr-xr-x 1 root root     47 Apr 28 07:18 .
drwxr-xr-x 1 root root     39 Apr 28 07:17 ..
-rw------- 1 root utmp      0 Nov 13  2020 btmp
drwxr-xr-x 1 fts3 root     46 May 27 09:17 fts3
drwxr-xr-x 1 fts3 fts3     26 Apr 28 07:18 fts3rest
-rw-r--r-- 1 root root    193 Nov 13  2020 grubby_prune_debug
drwx------ 1 root root     94 May 27 07:50 httpd
-rw-r--r-- 1 root root 292000 Apr 28 07:18 lastlog
-rw------- 1 root root  64000 Apr 28 07:18 tallylog
-rw-rw-r-- 1 root utmp      0 Nov 13  2020 wtmp
-rw------- 1 root root  14503 Apr 28 07:18 yum.log
```

It is created inside the `dev_fts_1` container when `test_tpc` gets executed inside the `dev_rucio_1` container. 

Since all tests run inside the `dev_rucio_1` container:

https://github.com/rucio/rucio/blob/8a0035701af3e79b905dbbed914607ea46983d20/.github/workflows/integration_tests.yml#L90-L92

the test environment cannot access the file system inside the `dev_fts_1` container. 

**Unless there is some way to access the filesystem of `dev_fts_1` container from inside of the `dev_rucio_1` container, I believe we cannot make this assertion in the `dev_rucio_1` container.** 

Since there is no straightforward way to do ^,  I export the artifact file from the 'dev_rucio_1' container that contains a string. This string is the location of the fts transfer log file inside the `dev_fts_1` container that was generated during the test.

The artifact file is then used in a subsequent step of integration_test.yml to check for the existence and contents of the fts transfer log file inside the `dev_fts_1` container:

https://github.com/rucio/rucio/blob/8a0035701af3e79b905dbbed914607ea46983d20/.github/workflows/integration_tests.yml#L93-L103 

The goal of the artifact mechanism is allow tests from 'dev_rucio_1' container to export information that can be used to perform additional checks on the states of the other containers and ensure that overall system level picture (xrd, fts, minio, ...) is looking good.

Let me know if we can get on a call and clarify this :)